File: python/strategy/Rifle.py
from Gun_interface import GunInterface

class Rifle(GunInterface):

    def __init__(self) -> None:
        super().__init__()
        self.magazine_size = 4
        self.bullet_count = 0

    def load(self):
        self.bullet_count = self.magazine_size
        print(".... magazine loaded.")

    def shoot(self, target):
        if self.bullet_count:
            self.bullet_count = self.bullet_count - 1
            print("BAM!")
            target.got_shoted()
        else:
            print("Out of bullets")
            self.load()
            

File: c++/strategy/src/Rifle.cpp
#include "Rifle.h"

using namespace std;

void Rifle::load() {
    _bullet_count = _magazine_size;
    cout << ".... magazine loaded." << endl;
}

void Rifle::shoot(TargetInterface* target) {
    if (_bullet_count) {
        _bullet_count--;
        cout << "BAM!" << endl;
        target->got_shoted();
    } else {
        cout << "Out of bullets." << endl;
        load();
    }
}